### PR TITLE
Refactor `Data.List.Relation.Binary.Permutation.*`, part I

### LINKED ISSUE #1354: Reimplement `Permutation.Propositional` in terms of `Permutation.Setoid`
Unlike other binary relations over lists, `Data.List.Relation.Binary.Permutation.(Setoid/Propositional)` are implemented separately which results in an awful lot of proof duplication. It also means that other modules (e.g. `Subset`) don't play nice when trying to transfer results from setoid equality to propositional equality. I've added the breaking tag as it remains to be seen if they can be unified in a backwards compatible manner.

### PR DESCRIPTION
**Pull request by jamesmckinna**
Part I of a project to fix #1354 

Highlights of the refactoring, cf. #2317 :
* Definitions don't change, so this is a 'conservative' first go, non-`breaking` for `v2.1`
* Additions to each of `Propositional` and `Setoid` and their `Properties` reconcile, more or less, the shared parts of the API
* Remove appeals to 'buggy' `PermutationReasoning` additional syntax #2319 
* General tidying up: `variable`s etc.
* Equivalence in `Propositional` of the relation and the  `Setoid (setoid _)` instance
* Refactor `Propositional.product-↭` in that spirit
* `Setoid` changes the proof of `↭-trans` in terms of new lemmas `↭-transˡ-≋` and `↭-transʳ-≋`; with them a sharpening of the analysis in `split` via a new lemma `↭-split` (purely structural; no WF-induction required)
* Move to `Homogeneous`, and deprecation in `Setoid`, of `steps`; towards removing legacy cruft now rendered obsolete by the above analysis
* (BUG?) `steps`  is deprecated in `Setoid`, but doesn't trigger the `WARNING_ON_USAGE` when the qualified-imported module `Permutation S` is `open`ed in `Properties`
* Final tidy up of `BagAndSetEquality` finishing off refactoring begun in #2321 

Not done, but could be:
* refactor `Setoid.Properties` into `Core` and derived
* ditto. for `Propositional`, and then rederived via the `Setoid (setoid _)` instance (eg even the `drop...` properties can be derived from those in `Setoid`) 

Won't do:
* move `Setoid` properties to `Homogeneous` (as in #2317) and knock-on refactoring of the `Core` properties etc.

### CODE REVIEW

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:201]: Looking at this diff, I don't understand where the definition of `drop-mid` has gone. Also is there a reason why we've now made `drop-mid′` publicly visible outside the module? You haven't listed this in the CHANGELOG so I'm guessing it's a  mistake?

Comment (MatthewDaggitt) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:201]: Oh weird, just spotted the reason. The proof is now mutually dependent.... Is this refactoring really necessary? Non-mutually dependent proofs seems more desirable?

File: src/Data/List/Relation/Binary/Permutation/Setoid.agda
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Binary/Permutation/Setoid.agda:47]: Again `↭-pointwise-≋`

File: src/Data/List/Relation/Binary/Permutation/Setoid.agda
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Binary/Permutation/Setoid.agda:144]: As mentioned above, not super keen on this deprecation. Can we simply re-export from `Homogeneous` for now?

File: src/Data/List/Relation/Binary/Permutation/Setoid.agda
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Binary/Permutation/Setoid.agda:93]: This is a big change in behaviour of `trans` and may break user's proofs? Regardless there should be documentation here explaining why we're using this non-obvious definition?

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:65]: I have to confess I don't understand the purpose of this? As far as I can see it saves two space characters and that's it?

File: src/Data/Nat/Primality/Factorisation.agda
Comment (MatthewDaggitt) [inline src/Data/Nat/Primality/Factorisation.agda:170]: Why are these changes necessary? If this PR is truly backwards compatible as you say, then can we remove all the changes in this file from this PR and make them in a separate one? Would be reassuring! (apart from the deprecated reasoning combinator!)

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:120]: By convention the `{xs}` argument should come first as it's mentioned first...

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]

**Comment by MatthewDaggitt** [comment]
Generally looks good though! The drastic reduction in proof length is pretty great!

File: src/Data/List/Relation/Binary/Permutation/Homogeneous.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Homogeneous.agda:66]: I can roll back on that, but interestingly, the warning doesn't seem to be generated when importing `Setoid` qualified in `Setoid.Properties`... (I raise that as a potential 'bug' above...)

Mind you, you should consider that induction on `steps` is *very* representation dependent, to the point of not even being invariant under the smart constructors such as `↭-trans`, and requiring eg that `↭-sym` actually preserve the shape of derivations on the nose... which isn't likely to survive any downstream representation change?

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:201]: Two or three 'reasons' for the refactoring:
* first I hadn't quite finished, alas, and was still doing minor tweaks; sorry for that, but now it's a helper function again...
* given the refactoring in `Setoid.Properties` of `↭-split`, I had hoped to be able to eliminate `drop-mid` in favour of `dropMiddleElement`... but then ran out of steam
* there's a very heavy reliance in the original on calling `drop-mid′` mostly with last two arguments `refl`, so it seemed worthwhile (for readability, if nothing else) to in-line those exactly in terms of `drop-mid` and hope that the termination checker could cope... and it can ;-)

File: src/Data/List/Relation/Binary/Permutation/Setoid.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid.agda:93]: Interesting!
... yes, the intentional behaviour does change, but exactly by analogy with the corresponding version in `Propositional`. But if you *really* want the intentional behaviour, you still have access to the vanilla `trans` constructor... is that (not) good enough?

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:65]: Yeah, it's maybe taking DRY too far, but it's purely local. Elsewhere in #2317 though, I ended up having to use the qualified version of `_++_` and that was getting very heavy...

Easy to inline though!

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:120]: Not always on top of the conventions, but yes, sure.

File: src/Data/Nat/Primality/Factorisation.agda
Comment (jamesmckinna) [reply src/Data/Nat/Primality/Factorisation.agda:170]: Well that's certainly true (separate PR), and maybe it's PR mission-creep, but once you start tackling `Propositional` permutation, tackling the (very modest) downstream knock-on opportunities seemed/seems worthwhile. Can revert, once again.

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:120]: DONE, and use o0f `variable`s means it need not even be quantified...

File: src/Data/List/Relation/Binary/Permutation/Propositional.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Propositional.agda:63]: A-ha! Will fix... DONE

File: src/Data/List/Relation/Binary/Permutation/Setoid.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid.agda:93]: DONE, as `↭-trans′`...

Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid.agda]: Mind you, a proof I couldn't be bothered to investigate is whether the `steps` properties hold for *this* version of 'smart' `trans`... which I think in fact should be provable, given the way pointwise equality is being pushed through the derivation tree to the leaves... but as I say, that's work I *really* don't want to do.

**Comment by jamesmckinna** [comment]
I think that's all of your review comments handled now! But leaving certain of the conversations unresolved until you're happy... ;-)

**Comment by jamesmckinna** [comment]
Regarding (the lack of extensionality in) `steps`: 
* intensionally, given a proof `p : [] ↭ []`, `steps p` will be some value well-defined by recursion on `p`
* extensionally, it feels *very* weird (to me) that we can't define a single value for this (eg 0, or 1, as currently defined as an instance of `refl`...)

Cf. #1113 Any such proof `p` must, in fact, hereditarily be some `trans` composition of proofs of the form `refl` (hence `↭-empty-inv` is not only provable, but has the proof it has), but each of those compositions will have a *different* `steps` count.

My analogy (apologies if I am belabouring the point) would be that of a `_++_`-based representation of `List` in which we counted `[] ++ xs` as having a different length to that of `xs`... no?

More generally, 'experience' suggests to me that definitions of inductively-defined relations which build in `trans` should generally be considered ... potentially harmful (eg it's a design pattern in type theory/PL to strive as far as possible for 'transitivity elimination') for precisely such 'reason's.

File: src/Data/List/Relation/Binary/Permutation/Homogeneous.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Homogeneous.agda]: I toned down the comment, but forgot to push that until now. Sorry!

File: CHANGELOG.md
Comment (jamesmckinna) [inline CHANGELOG.md:308]: It occurred to me that this deprecation, too, is premature. Will revert now.

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:499]: Premature, esp. in view of my comment elsewhere on preferring a (future!) version of this to take the *bundle* and not the *structure* as argument. But I should save the name for such downstream refactoring! Will now revert.

File: src/Data/List/Relation/Binary/Permutation/Setoid.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid.agda]: Out of interest, does the introduction of `↭-trans′`  break any of *your* proofs?

**Review by JacquesCarette — CHANGES_REQUESTED** [review]

File: src/Data/List/Relation/Binary/Permutation/Homogeneous.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Homogeneous.agda:67]: This is legacy code, and after @MatthewDaggitt 's (justified; correct) discouraging me from deprecating it, I don't propose to touch it as part of what is meant to be a conservative non-breaking PR...

File: src/Data/List/Relation/Binary/Permutation/Propositional.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Propositional.agda:91]: I don't think that's going to work for precisely the reasons that distinguish the use of *intensional* aspects of the definitions form their *extensional ones*. In particular, I don't think it will even hold without further binding in the specific constructor definitions for the two relations. 

In any case, as this PR is intended as the first part of a longer sequence intended to conclude with a resolution of both #1354 and #2311 , I'm inclined to say this review comment as mission-creep that's out-of-scope at this stage?

File: src/Data/List/Relation/Binary/Permutation/Homogeneous.agda
Comment (JacquesCarette) [reply src/Data/List/Relation/Binary/Permutation/Homogeneous.agda:67]: Sold.

File: src/Data/List/Relation/Binary/Permutation/Propositional.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Propositional.agda:133]: Again: legacy code.
All I have done here is encapsulate the permutation constructors and that of the underlying representing type `IsRelatedTo`, so that the appeal to `Reasoning.Syntax` is better encapsulated. 
NB. *no* connection here between `Homogeneous.steps` and the `step-*` combinators defined: the latter refers to steps of `Reasoning.Syntax`!

File: src/Data/List/Relation/Binary/Permutation/Propositional.agda
Comment (JacquesCarette) [reply src/Data/List/Relation/Binary/Permutation/Propositional.agda:91]: Ok.

File: src/Data/List/Relation/Binary/Permutation/Setoid.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid.agda:142]: Good catch!

File: src/Data/List/Relation/Binary/Permutation/Propositional.agda
Comment (JacquesCarette) [reply src/Data/List/Relation/Binary/Permutation/Propositional.agda:133]: Ok, happy to not devolve this into mission creep.

**Review by JacquesCarette — APPROVED** [review]

File: src/Data/List/Relation/Binary/Permutation/Setoid.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Setoid.agda]: Forgot to add the commentary you requested. Probably a bit verbose now!

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda]: I'll revert if you insist, but I do think an important part of `refactoring` is also to improve/streamline existing proofs... Please advise!

File: src/Data/List/Relation/Binary/Permutation/Homogeneous.agda
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Binary/Permutation/Homogeneous.agda:64]: This feels like a GitHub comment rather than a code comment. Maybe just a simple comment that says `Measures the number of constructors, can be useful for termination proofs` rather than a section heading?

**Review by MatthewDaggitt — COMMENTED** [review]
Other than that looks great!

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (MatthewDaggitt) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda]: Yup okay fair enough. The latest refactoring makes it clear that it's not necessary to make it mutually recursive so I'm happy with it.

File: src/Data/List/Relation/Binary/Permutation/Propositional.agda
Comment (gallais) [inline src/Data/List/Relation/Binary/Permutation/Propositional.agda:55]: As always I am going to be annoying and complain that I don't like prefixed
names because users can either use the short names or import the module
qualified if they need to manually disambiguate.

The same goes for `split` being renamed to `↭-split`  (although in that case
the name `split` is already pretty poor IMO).

Comment (jamesmckinna) [reply src/Data/List/Relation/Binary/Permutation/Propositional.agda:55]: @gallais I don't find the comment annoying!
What I find annoying is the the abstraction failure when explicitly exporting constructors from an inductive definition which should remain (externally, to clients) abstract, IMHO (cf. `z<s` in `Data.Nat.Base`, and `0<1+n` in `Data.Nat.Properties`, not that I/we are entirely consistent in enforcing this...)
Since this is also part of a greater project of refactoring `Permutation` towards completing #1761 / #1354 I wanted to make sure that in `Properties`, references to the constructors were abstracted on those lines before proceeding further.

As regards `↭-split`, this is *not* a renaming of `split`, but a significant generalisaiton which permits a much improved proof of `dropMiddleElement`... so a cognate name change is needed to relfect the 'bettre' typing!

**Comment by jamesmckinna** [comment]
Modulo @gallais 's comment, and my (attempted) rebuttal, I think that this is good to go @MatthewDaggitt unless there are review comments of yours that you are not happy have been dealt with...?

**Comment by jamesmckinna** [comment]
At some point, the merge conflict resolution (probably thanks to me) went wrong, so fixing things now...

**Review by MatthewDaggitt — APPROVED** [review]

**Review by JacquesCarette — APPROVED** [review]

**Comment by jamesmckinna** [comment]
Resolving latest round of merge conflicts exposed some issues with `CHANGELOG` (incl. a bug in that for #2418 ), which I've attempted to correctly rectify, but until this is merged, they'll probably go round and round in circles... sigh.

**Comment by jamesmckinna** [comment]
Hmmm.... fixing the merge conflict may need a bit more thought.

**Pull request merged**
